### PR TITLE
WOETH lending markets adjustements

### PR DESCRIPTION
**Pull request by sparrowDom**
## Code Change Checklist

To be completed before internal review begins:

- [ ]  The contract code is complete
- [ ]  Executable deployment file
- [ ]  Fork tests that test after the deployment file runs
- [ ]  Unit tests *if needed
- [ ]  The owner has done a [full checklist review](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) of the code + tests

Internal review:

- [ ] Two approvals by internal reviewers

## Deploy checklist

Two reviewers complete the following checklist:

```
- [ ] All deployed contracts are listed in the deploy PR's description
- [ ] Deployed contract's verified code (and all dependencies) match the code in master
- [ ] The transactions that interacted with the newly deployed contract match the deploy script.
- [ ] Governance proposal matches the deploy script
- [ ] Smoke tests pass after fork test execution of the governance proposal
```

### CODE REVIEW

**Comment by sparrowDom** [comment]
## Requirements

_What is the PR trying to do? Is this the right thing? Are there bugs in the requirements?_

## Easy Checks

#### Authentication

- [x] Never use tx.origin
- [x] Every external/public function is supposed to be externally accessible
- [x] Every external/public function has the correct authentication

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.
- [x] Contract is not vulnerable to being sent self destruct ETH

#### Cryptographic code

- [x] This contract code does not roll it's own crypto.
- [x] No signature checks without reverting on a 0x00 result.
- [x] No signed data could be used in a replay attack, on our contract or others.

#### Gas problems

- [ ] ~~Contracts with for loops must have either~~: _no loops_
    - [ ] ~~A way to remove items~~ _no loops_
    - [ ] ~~Can be upgraded to get unstuck~~ _no loops_
    - [ ] ~~Size can only controlled by admins~~ _no loops_
- [ ] ~~Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.~~ _no loops_

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0
- ~~[ ] All for loops use uint256~~ _no loops_

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding
- 🟠 Contract rounds in the protocols favor _it doesn't but it mimics what OUSD is doing using `mulTruncate` without rounding_
- [x] Contract does not have bugs from loosing rounding precision
- [x] Code correctly multiplies before division
- [x] Contract does not have bugs from zero or near zero amounts

#### Dependencies
- ~~[ ] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes.~~ _no new dependancies added_
- [ ] ~~If OpenZeppelin ACL roles are use review & enumerate all of them.~~
- [x] Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used.

#### External calls

- [ ] Contract addresses passed in are validated
- [ ] No unsafe external calls
- [ ] Reentrancy guards on all state changing functions
    - [ ] Still doesn't protect against external contracts changing the state of the world if they are called.
- [ ] No malicious behaviors
- [ ] Low level call() must require success.
- [ ] No slippage attacks (we need to validate expected tokens received)
- [ ] Oracles, one of:
  - [ ] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [ ] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting

#### Tests

  - [ ] Each publicly callable method has a test
  - [ ] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested
  - [ ] If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing.

#### Deploy

- [ ] Deployer permissions are removed after deploy

## Strategy Specific

_Remove this section if the code being reviewed is not a strategy._

#### Strategy checks

- [ ] Check balance cannot be manipulated up AND down by an attacker
- [ ] No read only reentrancy on downstream protocols during checkBalance
- [ ] All reward tokens are collected
- [ ] The harvester can sell all reward tokens
- [ ] No funds are left in the contract that should not be as a result of depositing or withdrawing
- [ ] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [ ] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation.
- [ ] WithdrawAll() cannot be MEV'd
- [ ] Strategist cannot steal funds

#### Downstream

- [ ] We have monitoring on all backend protocol's governances
- [ ] We have monitoring on a pauses in all downstream systems

## Thinking

#### Logic

  _Are there bugs in the logic?_

  - [ ] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).

#### Deployment Considerations

_Are there things that must be done on deploy, or in the wider ecosystem for this code to work. Are they done?_

#### Internal State

- What can be always said about relationships between stored state
- What must hold true about state before a function can run correctly (preconditions)
- What must hold true about the return or any changes to state after a function has run.

Does this code do that?

#### Attack

_What could the impacts of code failure in this code be._

_What conditions could cause this code to fail if they were not true._

_Does this code successfully block all attacks._

#### Flavor

_Could this code be simpler?_

_Could this code be less vulnerable to other code behaving weirdly?_

File: contracts/contracts/token/WOETH.sol [hunk 1]
```diff
@@ -6,17 +6,30 @@ import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
...
+ * credits per token in those 4 actions. That way WOETH can keep an accurate track of the OETH balance
+ * ignoring any unexpected transfers of OETH to this contract.
  */
 
 contract WOETH is ERC4626, Governable, Initializable {
     using SafeERC20 for IERC20;
+    using StableMath for uint256;
+    // doesn't need to be public, but convenient to be able to confirm the state on the mainnet
+    uint256 public oethCreditsHighres;
+    bool _oethCreditsInitialized;
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:32]: Missing an explicit visibility.

**Comment by DanielVF** [comment]
The core attack we are trying to stop is someone sending the OETH to the wOETH contract, causing the value of wOETH in OETH terms to go suddenly up.

It looks like totalAssets uses the amount of OETH held by the contract as one of two multipliers. totalAssets is in turn used to calculate the exchange ratio. If someone donates to the contract, one of these two multipliers goes up, and the donation has perfectly succeeded in increasing the value of each wOETH. This attack does not appear to be blocked at all?

Or am I missing something?

**Comment by DanielVF** [comment]
It also feels really scary that were are minting and burning using old ratios. That doesn't cause rektness?

File: contracts/contracts/token/WOETH.sol (hunk 1 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: thanks nice catch, corrected

**Comment by sparrowDom** [comment]
@DanielVF not completely sure what you mean. So the logic is that: 
- when the contract is initialized (on proposal execution) the internal credits of OETH are read and stored in WOETH. From that point forward the WOETH doesn't ask for internal credits OETH contract is holding but rather keeps its own accounting
- on each mint, deposit, withdraw, redeem the current credits per token are fetched from OETH (since we don't know if a rebase happened since the last time we queried it) and the accounting of credits in WOETH contracts are updated. Since only those 4 functions trigger the update of credits a normal transfer of OETH to WOETH contract has no effect. There is also [a test for this](https://github.com/OriginProtocol/origin-dollar/pull/2106/files#diff-1a55b6049ea6ba4898ff1fa84a24b20c6a6e2d20764957272f41ac0a6d01b177R109-R127) where you can see that after OETH transfer the WOETH account will still have a balance of 50 OETH and totalSupply of 0.

**Comment by DanielVF** [comment]
I misread the code and got the credits per token and the credits switched. Nevermind everything I said - I'll have to stare at it more!

File: contracts/contracts/token/WOETH.sol [hunk 2]
```diff
@@ -49,7 +81,161 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+
+    /**
+     * @dev This function converts requested OETH token amount to its underlying OETH
+     * credits value that is stored internally in OETH.sol and is required in order to
+     * be able to rebase.
+     *
+     * @param oethAmount Amount of OETH to be converted to OETH credits
+     * @return amount of OETH credits the OETH amount corresponds to
+     */
+    function _oethToOethCreditsHighres(uint256 oethAmount)
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:98]: Perhaps this would clearer if this method was named `_creditsPerAsset`, and the state variable `oethCreditsHighres` changed to `assetCredits`. 

In this way all the 4626 code gets to use clear variable names, and the OETH related stuff ends up just inside `_creditsPerAsset`.

File: contracts/contracts/token/WOETH.sol [hunk 3]
```diff
@@ -49,7 +81,161 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+         * of OETH.
+         */
+        return oethAmount.mulTruncate(creditsPerTokenHighres);
+    }
+
+    /** @dev See {IERC4262-totalAssets} */
+    function totalAssets() public view virtual override returns (uint256) {
+        (, uint256 creditsPerTokenHighres, ) = OETH(asset())
+            .creditsBalanceOfHighres(address(this));
+
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:122]: If we used in `credits` here as well from the call from `creditsBalanceOfHighres`, we would have ever piece of data already loaded inside this function to check that our actual assets was equal or greater than our expected assets.

**Comment by DanielVF** [comment]
Invariants I can think of:

- Assuming credits creditsPerTokenHighres in OETH only moves down, then woeth.convertToAssets() should only ever move up
- We should always be able to withdraw all WOETH to OETH. (This is harder to test, so another way of phrasing this is that we should always have enough OETH on hand for everyone to withdraw)
- OETH Donations to wOETH should not affect the convertToAssets() result
- WETH donations to oeth can affect the convertToAssets() result upwards after a rebase, and that's fine.

File: contracts/contracts/token/WOETH.sol [hunk 4]
```diff
@@ -6,17 +6,30 @@ import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
...
+ * token on mint, deposit, redeem, withdraw. The issue is that OETH is rebasing and OETH balances
+ * will change when the token rebases. For that reason we are tracking the WOETH contract credits and
+ * credits per token in those 4 actions. That way WOETH can keep an accurate track of the OETH balance
+ * ignoring any unexpected transfers of OETH to this contract.
  */
 
 contract WOETH is ERC4626, Governable, Initializable {
     using SafeERC20 for IERC20;
+    using StableMath for uint256;
+    // doesn't need to be public, but convenient to be able to confirm the state on the mainnet
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:30]: Can remove this comment - we default to public on most contract variables.

File: contracts/contracts/token/WOETH.sol (hunk 2 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:98]: good point thanks, renamed

File: contracts/contracts/token/WOETH.sol (hunk 3 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:122]: Hmm, that would be a good check that would detect possible mathematical irregularities (perhaps due to rounding). Though as you point out it would be weird to revert in a view function.

Monitoring sounds like a good place for it though. Have Grafana regularly query both credits values and check that the ones in OETH.sol are greater or equal to the ones in WOETH.sol.

Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:122]: Notion task added here: https://www.notion.so/originprotocol/WOETH-monitoring-sanity-check-695c73c0d0d04ef999a5242407d8a212

File: contracts/contracts/token/WOETH.sol (hunk 4 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: agreed

File: contracts/contracts/token/WOETH.sol [hunk 5]
```diff
@@ -31,6 +43,25 @@ contract WOETH is ERC4626, Governable, Initializable {
         OETH(address(asset())).rebaseOptIn();
     }
 
+    function initialize2() external onlyGovernor {
+        if (_oethCreditsInitialized) {
+            require(false, "Initialize2 already called");
+        }
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:49]: Why not just `require(!_oethCreditsInitialized, "Initialize2 already called");`?

File: contracts/contracts/token/WOETH.sol [hunk 6]
```diff
@@ -49,7 +80,161 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+
+    /**
+     * @dev This function converts requested OETH token amount to its underlying OETH
+     * credits value that is stored internally in OETH.sol and is required in order to
+     * be able to rebase.
+     *
+     * @param oethAmount Amount of OETH to be converted to OETH credits
+     * @return amount of OETH credits the OETH amount corresponds to
+     */
+    function _creditsPerAsset(uint256 oethAmount)
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:97]: nit: This function name confuses me a bit. When I see `per asset`, I assume it's gonna return credits for 1 OETH. But it takes in a `oethAmount` and multiplies the credits by it. So, it sounds more like `creditsForAssetAmount` than `creditsPerAsset`

File: contracts/contracts/token/WOETH.sol [hunk 7]
```diff
@@ -49,7 +80,161 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+            caller,
+            address(this),
+            assets
+        );
+        _mint(receiver, shares);
+        oethCreditsHighres += _creditsPerAsset(assets);
+
+        emit Deposit(caller, receiver, assets, shares);
+
+        return shares;
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:153]: Could this be simplified into just (to avoid code-duplication):
```
shares = super.deposit(assets, receiver);
oethCreditsHighres += _creditsPerAsset(assets);
```

Comment (shahthepro) [reply contracts/contracts/token/WOETH.sol:153]: Same with other overridden functions as well

File: contracts/contracts/token/WOETH.sol [hunk 8]
```diff
@@ -49,7 +80,161 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+        uint256 shares = previewWithdraw(assets);
+
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // if _asset is ERC777, transfer can call reenter AFTER the transfer happens through
+        // the tokensReceived hook, so we need to transfer after we burn to keep the invariants.
+        _burn(owner, shares);
+        oethCreditsHighres -= _creditsPerAsset(assets);
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:205]: Haven't tested it yet but I wonder if we should compute the credits before burn? Let me see if it makes any difference or if it's breakable

Comment (shahthepro) [reply contracts/contracts/token/WOETH.sol:205]: I tried running some tests. Change in order didn't affect the numbers at all

File: contracts/contracts/token/WOETH.sol [hunk 9]
```diff
@@ -49,7 +80,161 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+        _mint(receiver, shares);
+        oethCreditsHighres += _creditsPerAsset(assets);
+
+        emit Deposit(caller, receiver, assets, shares);
+
+        return shares;
+    }
+
+    /** @dev See {IERC4262-mint} */
+    function mint(uint256 shares, address receiver)
```
Comment (pandadefi) [inline contracts/contracts/token/WOETH.sol:157]: You could save on some code complexity calling deposit from the mint function to prevent having most of the same code twice.

**Comment by pandadefi** [comment]
A donation to increase the OETH price remains possible via the OETH contract but will require significant funds to increase the entire OETH vault value.

File: contracts/contracts/token/WOETH.sol (hunk 5 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: good point: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/057469c08067ff67e31b8e1f524b61e249787499

File: contracts/contracts/token/WOETH.sol (hunk 9 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:135]: thanks, yes agreed. done here: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/1a6a16fc4af1f115f5c84a331df821da4b551d47

File: contracts/contracts/token/WOETH.sol (hunk 6 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:95]: yup agreed: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/1a6a16fc4af1f115f5c84a331df821da4b551d47

File: contracts/contracts/token/WOETH.sol (hunk 7 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: yup good point. done here: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/1a6a16fc4af1f115f5c84a331df821da4b551d47

File: contracts/contracts/token/WOETH.sol (hunk 6 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: good point. I have also renamed assets & shares since in the context of WOETH those can only be OETH & WOETH: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/2223600a1d8b8bf23c940a7debe124ddb1e4f44d

**Comment by sparrowDom** [comment]
@pandadefi yes the donation attack on OETH with donating WETH to the Vault and calling `rebase` is ok. This is the way OETH increases its value and I can't think of a way how it could be exploited

File: contracts/contracts/token/WOETH.sol [hunk 10]
```diff
@@ -6,17 +6,29 @@ import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
...
+ * will change when the token rebases. For that reason we are tracking the WOETH contract credits and
+ * credits per token in those 4 actions. That way WOETH can keep an accurate track of the OETH balance
+ * ignoring any unexpected transfers of OETH to this contract.
  */
 
 contract WOETH is ERC4626, Governable, Initializable {
     using SafeERC20 for IERC20;
+    using StableMath for uint256;
+    uint256 public oethCreditsHighres;
+    bool private _oethCreditsInitialized;
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:31]: `WOETHBase` inherits `WOETH`, so might as well add some gaps here now

File: contracts/contracts/token/WOETH.sol [hunk 11]
```diff
@@ -49,7 +78,84 @@ contract WOETH is ERC4626, Governable, Initializable {
         external
         onlyGovernor
     {
+        //@dev TODO: we could implement a feature where if anyone sends OETH direclty to
+        // the contract, that we can let the governor transfer the excess of the token.
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:82]: nit: perhaps we could just treat any donation as "yield"?

File: contracts/contracts/token/WOETH.sol [hunk 12]
```diff
@@ -49,7 +78,84 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+    function _oethToCredits(uint256 oethAmount) internal returns (uint256) {
+        (, uint256 creditsPerTokenHighres, ) = OETH(asset())
+            .creditsBalanceOfHighres(address(this));
+
+        /**
+         * Multiplying OETH amount with the creditsPerTokenHighres is exactly the math that
+         * is internally being done in OETH:
+         */
+        // solhint-disable-next-line max-line-length
+        /** https://github.com/OriginProtocol/origin-dollar/blob/2314cccf2933f5c1f76a6549c1f5c9cc935b6f05/contracts/contracts/token/OUSD.sol#L242-L249
```
Comment (shahthepro) [inline contracts/contracts/token/WOETH.sol:104]: nit: Points to older code (pre-yield-delegation). Would be nice to update it to point to latest commit hash on master

File: contracts/contracts/token/WOETH.sol (hunk 10 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:31]: good point, done

File: contracts/contracts/token/WOETH.sol [hunk 13]
```diff
@@ -31,11 +43,40 @@ contract WOETH is ERC4626, Governable, Initializable {
         OETH(address(asset())).rebaseOptIn();
     }
 
-    function name() public view virtual override returns (string memory) {
+    function initialize2() external onlyGovernor {
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:46]: Perhaps we should have `initialize()` call `initialize2()`. This way new contract deploys don't need to call both, and we are less likely to make a mistake.

File: contracts/contracts/token/WOETH.sol (hunk 10 above)
Comment (DanielVF) [reply contracts/contracts/token/WOETH.sol:32]: I don't see the gaps?

File: contracts/contracts/token/WOETH.sol [hunk 14]
```diff
@@ -5,24 +5,36 @@ import { ERC4626 } from "../../lib/openzeppelin/contracts/token/ERC20/extensions
...
+ * If OETH weren't rebasing this could be achieved by solely tracking the ERC20 transfers of the OETH
+ * token on mint, deposit, redeem, withdraw. The issue is that OETH is rebasing and OETH balances
+ * will change when the token rebases. For that reason we are tracking the WOETH contract credits and
+ * credits per token in those 4 actions. That way WOETH can keep an accurate track of the OETH balance
+ * ignoring any unexpected transfers of OETH to this contract.
  */
 
 contract WOETH is ERC4626, Governable, Initializable {
     using SafeERC20 for IERC20;
+    using StableMath for uint256;
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:30]: This contract code is way cleaner than the last time I looked at it. Good job!

File: contracts/contracts/token/WOETH.sol [hunk 15]
```diff
@@ -49,7 +90,84 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+        woethAmount = super.withdraw(oethAmount, receiver, owner);
+        oethCreditsHighres -= _oethToCredits(oethAmount);
+    }
+
+    /** @dev See {IERC4262-redeem} */
+    function redeem(
+        uint256 woethAmount,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256 oethAmount) {
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:169]: This contract does not currently compile. This should be fixed.

Also, although this is not the actual compile error, I'm wondering if these `virtual`s in these methods are wrong, since I think we want these functions callable without needing to be overridden by a child class.

File: contracts/contracts/token/WOETH.sol (hunk 11 above)
Comment (DanielVF) [reply contracts/contracts/token/WOETH.sol:94]: While we shouldn't treat donations as instant yield (that's what this contract is trying to get away from), I do think we should build in a separate governor method to collect donated funds that are in excess of the backing funds.

File: contracts/contracts/token/WOETH.sol [hunk 16]
```diff
@@ -49,7 +90,84 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+    /**
+     * @dev This function converts requested OETH token amount to its underlying OETH
+     * credits value that is stored internally in OETH.sol and is required in order to
+     * be able to rebase.
+     *
+     * @param oethAmount Amount of OETH to be converted to OETH credits
+     * @return amount of OETH credits the OETH amount corresponds to
+     */
+    function _oethToCredits(uint256 oethAmount) internal returns (uint256) {
+        (, uint256 creditsPerTokenHighres, ) = OETH(asset())
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:108]: Both `_oethToCredits()` and `totalAssets()` call `oeth.creditsBalanceOfHighres()` to get the `creditsPerToken`, discarding the other returned values.

I think it makes more sense to use the simpler `oeth.rebasingCreditsPerTokenHighres()`. 

There's two scenarios here:

1) This w token is correctly marked as rebasing. In this case, `oeth.rebasingCreditsPerTokenHighres()` will return the same value, be simpler, return only what we need, and cost less gas.

2) Governance messes up the world in a bad way and turns off yield to the contract. The current call will immediately return 1e18 instead of 1e27ish, making for a really really wrong totalAssets, and thus really really wrong conversion rate. However, if `oeth.rebasingCreditsPerTokenHighres()` is used we'll only get a gradual drift off the correct value as expected yield does not come in.

In both cases the behavior of `oeth.rebasingCreditsPerTokenHighres()` seems better.

File: contracts/contracts/token/WOETH.sol [hunk 17]
```diff
@@ -49,7 +90,84 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+    }
+
+    /** @dev See {IERC4262-redeem} */
+    function redeem(
+        uint256 woethAmount,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256 oethAmount) {
+        oethAmount = super.redeem(woethAmount, receiver, owner);
+        oethCreditsHighres -= _oethToCredits(oethAmount);
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:171]: This is just a mathematical nit, and perhaps the code is okay without this, but in general, if you are depositing/plus-ing and withdrawing/minus-ing using the same conversion function, you are almost certainly rounding the wrong direction in one of them. It's possible we need two `_oethToCredits`, one that round up, and one that rounds down.

File: contracts/contracts/token/WOETH.sol (hunk 10 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:32]: Forgot to push remotely: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/be6d4734ec071df5ba9801fbc85bdfacb69c2007

File: contracts/contracts/token/WOETH.sol (hunk 12 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: nice catch thanks: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/c38a3c659a62fd81807dadaca0151b629772deae

File: contracts/contracts/token/WOETH.sol (hunk 15 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: Yes you are right these virtual keywords are not needed. We can add it in the future if need be: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/5ff5c5dfd0f3e81d386b6396894b954dfcf5f78b

File: contracts/contracts/token/WOETH.sol (hunk 16 above)
Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol]: Great comment and great points thanks: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/3ef22194932715aac4bc2dca3d782c0769d2b420

File: contracts/contracts/token/WOETH.sol [hunk 18]
```diff
@@ -62,7 +99,81 @@ contract WOETH is ERC4626, Governable, Initializable {
...
+         * up / down depending on what favours this contract
+         */
+        // solhint-disable-next-line max-line-length
+        /** https://github.com/OriginProtocol/origin-dollar/blob/c02572bd1c06eb3c2652c8692e52144be2efa741/contracts/contracts/token/OUSD.sol#L498
+         */
+        return (oethAmount * creditsPerTokenHighres + (roundUp ? 1e18 - 1 : 0)) / 1e18;
+    }
+
+    /** @dev See {IERC4262-totalAssets} */
+    function totalAssets() public view override returns (uint256) {
```
Comment (DanielVF) [inline contracts/contracts/token/WOETH.sol:133]: 🔴 `rebasingCreditsPerToken` has the property that the smaller it is, the more money a user has.

Default solidity division results in rounding down. Therefore this `totalAssets()` method is rounding up the amount of money in contract. Therefore it is very easy for the contract to be a rounding error on the wrong side of solvent.

Comment (DanielVF) [reply contracts/contracts/token/WOETH.sol:133]: That said, we should be careful to fuzz test the behavior differences. There's no perfect answer here, and so we want to make sure that erroring on the solvent side is correct.

Comment (DanielVF) [reply contracts/contracts/token/WOETH.sol:133]: This is all it takes for the contract to report insolvancy with the current code.

image

Comment (sparrowDom) [reply contracts/contracts/token/WOETH.sol:107]: a different approach has been made where the credits are read before the change and after from the OETH contract. This way the error in WOETH calculated internal credits and the one in OETH contract shouldn't diverge and cause potential conversion rate inaccuracies: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/c005824d9062cef9705fb02aaa2114ee64241362
